rir/src/ir/Compiler.cpp

```diff
@@ -145,6 +145,141 @@ Code* compilePromise(CompilerContext& ctx, SEXP exp);
+    cs << BC::brfalse(nextBranch);
+
+    compileBody();
+    cs << BC::pop() << BC::br(loopBranch) << nextBranch;
+
+    if (ctx.loopNeedsContext()) {
+        cs << BC::endloop();
+    } else {
+        cs.remove(beginLoopPos);
+    }
+
+    cs << BC::push(R_NilValue) << BC::invisible();
+
+    ctx.popLoop();
+}
+
+bool compileSimpleFor(CompilerContext& ctx, SEXP sym, SEXP seq, SEXP body) {
+    Match(seq) {
+        Case(LANGSXP, fun, argsSexp) {
+            RList args(argsSexp);
+            if (fun != symbol::Colon || args.length() != 2) {
+                return false;
+            }
+
+            SEXP start = args[0];
+            SEXP end = args[1];
+            if (TYPEOF(start) != INTSXP && TYPEOF(start) != LGLSXP &&
+                (TYPEOF(start) != REALSXP || XLENGTH(start) == 0 ||
+                 *REAL(start) != (int)*REAL(start))) {
+                return false;
+            }
+
+            // for(i in m:n) {
+            //   ...
+            // }
+            // =>
+            // i' <- m
+            // n' <- n
+            // if (i' > n') {
+            //   n' <- ceil(n') - 1
+            //   while (i' > n') {
+            //     i <- i'
+            //     i' <- i' - 1
+            //     ...
+            //   }
+            // } else {
+            //   n' <- floor(n') + 1
+            //   while (i' < n') {
+            //     i <- i'
+            //     i' <- i' + 1
+            //     ...
+            //   }
+            // }
+
+            CodeStream& cs = ctx.cs();
+            BC::Label fwdBranch = cs.mkLabel();
+            BC::Label endBranch = cs.mkLabel();
+
+            // i' <- m
+            compileExpr(ctx, start);
+            cs << BC::floor();
+            // n' <- n
+            compileExpr(ctx, end);
+            cs << BC::ensureNamed();
+            // if (i' > n')
+            cs << BC::dup2() << BC::gt();
+            cs.addSrc(R_NilValue);
+            cs << BC::brfalse(fwdBranch);
+            // {
+            // n' <- ceil(n') - 1
+            cs << BC::ceil() << BC::dec();
+            // while
+            compileWhile(ctx,
+                         [&cs]() {
+                             // (i' > n')
+                             cs << BC::dup2() << BC::gt();
+                             cs.addSrc(R_NilValue);
+                         },
+                         [&ctx, &cs, &sym, &body]() {
+                             // {
+                             // i <- i'
+                             cs << BC::pull(1) << BC::stvar(sym);
+                             // i' <- i' - 1
+                             cs << BC::swap() << BC::dec() << BC::swap();
+                             // ...
+                             compileExpr(ctx, body);
+                             // }
+                         });
+            // } else {
+            cs << BC::br(endBranch) << fwdBranch;
+            // n' <- floor(n') + 1
+            cs << BC::floor() << BC::inc();
+            // while
```

Comment:
missing ensure named

Reply:
inc should not be used here, since you don't modify in-place. use just `push 1; add`

Reply:
Similar to above